src/main/java/org/traccar/api/security/OpenIDProvider.java

```diff
@@ -0,0 +1,245 @@
+                .state(new State())
+                .nonce(new Nonce())
+                .build();
+
+        return request.toURI();
+    }
+
+    private OIDCTokenResponse getToken(AuthorizationCode code) {
+        // Credentials to authenticate us to the token endpoint
+        ClientAuthentication clientAuth = new ClientSecretBasic(this.clientId, this.clientSecret);
+        AuthorizationGrant codeGrant = new AuthorizationCodeGrant(code, this.callbackUrl);
+
+        TokenRequest request = new TokenRequest(this.tokenUrl, clientAuth, codeGrant);
+        TokenResponse tokenResponse;
+
+        try {
+            HTTPResponse tokenReq = request.toHTTPRequest().send();
+            tokenResponse = OIDCTokenResponseParser.parse(tokenReq);
+            if (!tokenResponse.indicatesSuccess()) {
+                return null;
+            }
+
+            return (OIDCTokenResponse) tokenResponse.toSuccessResponse();
+        } catch (IOException e) {
+            return null;
+        } catch (ParseException e) {
+            return null;
+        }
+    }
+
+    private AuthorizationCode parseCallback(URI requri) throws WebApplicationException {
+        AuthorizationResponse response;
+
+        try {
+            response = AuthorizationResponse.parse(requri);
+        } catch (ParseException e) {
+            return null;
+        }
+
+        if (!response.indicatesSuccess()) {
+            AuthorizationErrorResponse error = response.toErrorResponse();
+            throw new WebApplicationException(Response.status(403).entity(error.getErrorObject().getDescription()).build());
+        }
+
+        return response.toSuccessResponse().getAuthorizationCode();
+    }
+
+    private UserInfo getUserInfo(BearerAccessToken token) {
+        UserInfoResponse userInfoResponse;
+
+        try {
+            HTTPResponse httpResponse = new UserInfoRequest(this.userInfoUrl, token)
+                    .toHTTPRequest()
+                    .send();
+
+            userInfoResponse = UserInfoResponse.parse(httpResponse);
+        } catch (IOException e) {
+            return null;
+        } catch (ParseException e) {
+            return null;
+        }
+
+        if (!userInfoResponse.indicatesSuccess()) {
+            // User info request failed - usually from expiring
+            return null;
+        }
+
+        return userInfoResponse.toSuccessResponse().getUserInfo();
+    }
+
+    private User createUser(String name, String email, Boolean administrator) throws StorageException {
+        User user = new User();
+
+        user.setName(name);
+        user.setEmail(email);
+        user.setFixedEmail(true);
+        user.setDeviceLimit(this.config.getInteger(Keys.USERS_DEFAULT_DEVICE_LIMIT));
+
+        int expirationDays = this.config.getInteger(Keys.USERS_DEFAULT_EXPIRATION_DAYS);
+
+        if (expirationDays > 0) {
+            user.setExpirationTime(new Date(System.currentTimeMillis() + expirationDays * 86400000L));
+        }
```

Comment:
We should probably extract this logic somewhere to avoid duplication. Maybe `UserUtil`?

Reply:
Is this needed for our user creation? Doesn't look like we do it for LDAP at the moment, but we do when manually creating a user.

Reply:
This is probably needed for LDAP as well. I think it was just never implemented there.